### PR TITLE
Create datasets from Kobo

### PR DESCRIPTION
**Pull request by avdata99**
Related to #625 
Continue (and to be merged after) #635 

**Still work in progress**
This PR allow import datasets from KoBo resources
Will be continued with #626 to create actual resources

### CODE REVIEW

File: ckanext/unhcr/templates/kobo/surveys.html [hunk 1]
```diff
@@ -29,30 +29,20 @@ <h1>My KoBo Surveys</h1>
...
-                        <option value="{{ org.id }}">{{ org.name }}</option>
-                        {% endfor %}
-                    </select>
-                </div>
 
-            <!-- <button #625
+    <form method="POST" action="{% url_for 'unhcr_kobo.kobo_import_form' %}">
+        <input type="hidden" value="{{ survey.uid }}" name="kobo_asset_id" />
+        <div class="form-actions">
+            <button #625
```
Comment (amercader) [inline ckanext/unhcr/templates/kobo/surveys.html:36]:
```suggestion
            <button
```

File: ckanext/unhcr/templates/kobo/import-form.html [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    <select id="field-organizations" name="owner_org" data-module="autocomplete" required="required">
+      {% for organization in h.user_orgs('create_dataset') %}
+        <option value="{{ organization.id }}">{{ organization.display_name }}</option>
+      {% endfor %}
+    </select>
+  </div>
+</div>
+
+{% set schema = h.scheming_get_schema('dataset', 'dataset') %}
+{% set fields = h.scheming_get_schema('dataset', 'dataset') %}
```
Comment (amercader) [inline ckanext/unhcr/templates/kobo/import-form.html:69]: Duplicated?

File: ckanext/unhcr/templates/kobo/import-form.html [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      {% for key, error in errors.items() %}
+        <li data-field-label="{{ key }}">{% if key %}{{ key }}: {% endif %}{{ error }}</li>
+      {% endfor %}
+    </ul>
+  </div>
+{% endif %}
+
+{% endblock %}
+
+<form method="POST" action="{% url_for 'unhcr_kobo.kobo_import' %}">
```
Comment (amercader) [inline ckanext/unhcr/templates/kobo/import-form.html:24]: I'm getting this error when trying to create a dataset. Maybe there is this field missing?
Screenshot 2021-07-28 at 12-14-22 Import dataset from KoBo - RIDL

File: ckanext/unhcr/templates/kobo/import-form.html [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+<h2>Resource</h2>
+<div>
+  <p>
+    You are about to create a dataset with the survey data as resource.
+  </p>
+</div>
+<div class="form-group control-medium control-select">
+  <span title="This field is required" class="control-required">*</span>
+  <label for="field-data_version">Resource version name</label>
+  <input type="text" name="data_version" value="{{ data.data_version }}" required="required" placeholder="resource version name e.g. v1" id="field-data_version" size=50 maxlength="50">
```
Comment (amercader) [inline ckanext/unhcr/templates/kobo/import-form.html:36]:
```suggestion
  <div class="controls">
    <input type="text" name="data_version" value="{{ data.data_version }}" required="required" placeholder="resource version name e.g. v1" id="field-data_version" class="form-control" />
 </div>
```

Comment (amercader) [inline ckanext/unhcr/templates/kobo/import-form.html:36]: Can you tell me more about this field? Is this something present on the KoBo metadata? Is it some random name that users need to fill in? Where does it get stored in CKAN?

**Review by amercader — COMMENTED** [review]
@avdata99 I know this is a WIP but just added some early comments

File: ckanext/unhcr/templates/kobo/import-form.html (hunk 2 above)
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/import-form.html]: yes, fixed

File: ckanext/unhcr/templates/kobo/import-form.html (hunk 4 above)
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/import-form.html]: This is a [resource level required field](https://github.com/okfn/ckanext-unhcr/blob/master/ckanext/unhcr/schemas/dataset.json#L838-L842). My plan is to use `this-version-value__CSV` and `this-version-value__JSON` (or something like this) for each resource we create here.

Comment (amercader) [reply ckanext/unhcr/templates/kobo/import-form.html]: Ok, tbh we can probably generate this ourselves or at least present a default value for users, but let's discuss with Mariann.

What about the rest of mandatory resource fields (dates, Personally Identifiable, etc). My understanding from https://github.com/okfn/ckanext-unhcr/issues/626 is that after creating the dataset, users will be redirected to the standard resource form but with an extra "KoBo Data file" option next to the upload and link buttons, so they can add the rest of the Metadata

File: ckanext/unhcr/actions.py [hunk 5]
```diff
@@ -66,6 +67,22 @@ def package_update(up_func, context, data_dict):
...
+    """ Create resources for KoBo assets """
+
+    # Create dataset
+    dataset = up_func(context, data_dict)
+
+    kobo_asset_id = data_dict.get('kobo_asset_id')
+    if kobo_asset_id:
+        # create basic resources
+        kd = KoboDataset(kobo_asset_id)
+        kd.initialize_package(context, dataset)
```
Comment (avdata99) [inline ckanext/unhcr/actions.py:81]: To work at #626 and create basic resources from KoBo

File: ckanext/unhcr/blueprints/kobo.py [hunk 6]
```diff
@@ -90,12 +95,38 @@ def kobo_surveys():
         toolkit.h.flash_error(message)
         return toolkit.redirect_to('unhcr_kobo.index')
 
+    # connect packages if already been imported
+    for survey in surveys:
+        survey['ridl_package'] = KoboDataset(kobo_asset_id=survey['uid']).get_package()
```
Comment (avdata99) [inline ckanext/unhcr/blueprints/kobo.py:100]: Notify KoBo assets (surveys) if a dataset already has been created here

File: ckanext/unhcr/templates/package/snippets/package_form.html [hunk 7]
```diff
@@ -1,18 +1,34 @@
...
 
-{# This provides a full page that renders a form for adding a dataset. It can
-then itself be extended to add/remove blocks of functionality. #}
+{# This file extends from core package-form to add/remove functionality. #}
+
+{# Check if we are importing assets from KoBo #}
+{% set kobo_asset_id = request.args.get('kobo_asset_id') %}
+{% if kobo_asset_id %}
+    {% set kobo_pkg, errors = h.get_kobo_initial_dataset(kobo_asset_id) %}
+    {% set x = data.update(kobo_pkg) %}
```
Comment (pdelboca) [inline ckanext/unhcr/templates/package/snippets/package_form.html:10]: FYI: CKAN has the `do` extension of jinja2 that allows you to update dicts using:

```jinja
 {% do data.update(kobo_pkg) %}
```

and avoid an extra dummy variable.

File: ckanext/unhcr/templates/package/snippets/import-kobo-stages.html [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  </li>
+  <li class="last {{ s2 }}">
+    {% if s2 != 'complete' %}
+      <span class="highlight">{{ _('Describe KoBo resources') }}</span>
+    {% else %}
+      {% if s1 == 'active' %}
+        {# stage 1 #}
+        <button class="highlight" name="save" value="go-resources" type="submit">{{ _('Describe KoBo resources') }}</button>
+      {% else %}
+        {% link_for _('Describe KoBo resources'), named_route='dataset.new', class_="highlight" %}
```
Comment (pdelboca) [inline ckanext/unhcr/templates/package/snippets/import-kobo-stages.html:22]: CKAN is moving away from snippets that renders HTML so I will suggest to change this to HTML sintax as you did in previous template.

**Review by pdelboca — COMMENTED** [review]
I added some non-mandatory comments.

File: ckanext/unhcr/templates/package/snippets/import-kobo-stages.html (hunk 8 above)
Comment (avdata99) [reply ckanext/unhcr/templates/package/snippets/import-kobo-stages.html:22]: This is the same at [CKAN core](https://github.com/ckan/ckan/blob/2.9/ckan/templates/package/snippets/stages.html#L33-L36). I don't get it. What should the change be @pdelboca ?

**Pull request merged**